File: src/market-tracker-android/app/src/main/java/pt/isel/markettracker/http/models/product/ProductOutputModel.kt
package pt.isel.markettracker.http.models.product

import pt.isel.markettracker.domain.model.market.inventory.Category
import pt.isel.markettracker.domain.model.market.inventory.Brand
import pt.isel.markettracker.domain.model.market.inventory.ProductUnit
import pt.isel.markettracker.domain.model.market.inventory.product.Product

data class ProductOutputModel(
    val id: String,
    val name: String,
    val imageUrl: String,
    val quantity: Int,
    val unit: String,
    val rating: Double,
    val brand: Brand,
    val category: Category
) {
    fun toProduct() = Product(
        id,
        name,
        imageUrl,
        quantity,
        ProductUnit.fromTitle(unit),
        rating,
        brand,
        category
    )
}

File: src/market-tracker-webapi/Application/Http/Models/Schemas/Market/Inventory/Product/ProductOutputModel.cs
using market_tracker_webapi.Application.Domain.Schemas.Market.Inventory;
using market_tracker_webapi.Application.Domain.Schemas.Market.Inventory.Product;
using market_tracker_webapi.Application.Http.Models.Schemas.Market.Inventory.Category;

namespace market_tracker_webapi.Application.Http.Models.Schemas.Market.Inventory.Product;

using Product = Domain.Schemas.Market.Inventory.Product.Product;

public record ProductOutputModel(
    string Id,
    string Name,
    string ImageUrl,
    int Quantity,
    string Unit,
    double Rating,
    Brand Brand,
    CategoryOutputModel Category
);

public record ProductItemOutputModel(string ProductId, string Name, string ImageUrl, string BrandName);

public static class ProductInfoOutputModelMapper
{
    public static ProductOutputModel ToOutputModel(this Product product)
    {
        return new ProductOutputModel(
            product.Id.Value,
            product.Name,
            product.ImageUrl,
            product.Quantity,
            product.Unit.GetUnitName(),
            product.Rating,
            product.Brand,
            product.Category.ToOutputModel()
        );
    }
    
    public static ProductItemOutputModel ToProductItemOutputModel(this Product product)
    {
        return new ProductItemOutputModel(
            product.Id.Value,
            product.Name,
            product.ImageUrl,
            product.Brand.Name
        );
    }
}
